dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -25,8 +36,131 @@ namespace ErrorCodes
 {
 extern const int BAD_ARGUMENTS;
 extern const int NOT_IMPLEMENTED;
+extern const int CANNOT_INSERT_NULL_IN_ORDINARY_COLUMN;
 } // namespace ErrorCodes
 
+namespace
+{
+template <typename T>
+consteval bool checkIfSimpleNumericType()
+{
+    return std::is_integral_v<T> || std::is_floating_point_v<T>;
+}
+
+template <typename T>
+struct DecimalWithScale
+{
+    DecimalWithScale(T decimal_, Int64 scale_)
+        : decimal(decimal_)
+        , scale(scale_)
+    {
+        if constexpr (!IsDecimal<T>)
+            throw Exception("DecimalWithScale only accepts Decimal type.");
+    }
+
+    T decimal; // This T should be decimal type
+    Int64 scale;
+};
+
+using DecimalWithScale32 = DecimalWithScale<Decimal32>;
+using DecimalWithScale64 = DecimalWithScale<Decimal64>;
+using DecimalWithScale128 = DecimalWithScale<Decimal128>;
+using DecimalWithScale256 = DecimalWithScale<Decimal256>;
+
+template <typename T>
+consteval bool checkIfDecimalWithScaleType()
+{
+    return std::is_same_v<T, DecimalWithScale32> || std::is_same_v<T, DecimalWithScale64> || std::is_same_v<T, DecimalWithScale128> || std::is_same_v<T, DecimalWithScale256>;
+}
+
+// When T is Decimal, we should convert it to DecimalWithScale type
+// as we need scale value when executing the comparison operation.
+template <typename T>
+struct ActualCmpDataType
+{
+    using Type = std::conditional_t<checkIfSimpleNumericType<T>(), T, DecimalWithScale<T>>;
+};
+
+template <typename T>
+typename ActualCmpDataType<T>::Type getValue(const ColumnPtr & col_ptr, size_t idx)
+{
+    if constexpr (checkIfSimpleNumericType<T>())
+    {
+        const auto * col = static_cast<const ColumnVector<T> *>(&(*col_ptr));
+        const typename ColumnVector<T>::Container & data = col->getData();
+        return data[idx];
+    }
+    else if (IsDecimal<T>)
+    {
+        const auto * col = static_cast<const ColumnDecimal<T> *>(&(*col_ptr));
```

Comment:
Could it be ColumnConst?
Does Decimal appear here?

Reply:
> Could it be ColumnConst? Does Decimal appear here?

1. It's impossible to be const column here
2. This if branch is especially used for handling Decimal.

Reply:
> > Could it be ColumnConst? Does Decimal appear here?
> 
> 1. It's impossible to be const column here
> 2. This if branch is especially used for handling Decimal, so the Decimal will appear here.

Reply:
Why must it not be ColumnConst, doesn't this depend on the input?
What if the operator below window outputs ColumnConst?

Reply:
> Why must it not be ColumnConst, doesn't this depend on the input? What if the operator below window outputs ColumnConst?

Yes, it depend on the input, but tidb will ensure that the input column that user designate is not const.
```sql
select *, first_value(v) over (partition by p order by 1 range between 1 preceding and 1 following) over a from a; // Error, order by 1, 1 is a const column.
```

Reply:
> what if `select *, first_value(v) over (partition by p order by a range between b preceding and c following) over a from (select 1 as a, 2 as b, 3 as c, v from table); `?
I will add comment and test for it.